File: scripts/finetune/portions.py
"""Python port of src/lib/portions.ts convert_portion.

Grams from NIST/FDA/USDA household weights. Calories from USDA per 100 g only.
"""

from __future__ import annotations

import re
from typing import Any

UNIT_ALIAS = {
    "gram": "g",
    "grams": "g",
    "kilogram": "kg",
    "kilograms": "kg",
    "ounce": "oz",
    "ounces": "oz",
    "pound": "lb",
    "pounds": "lb",
    "lbs": "lb",
    "milliliter": "ml",
    "milliliters": "ml",
    "millilitre": "ml",
    "millilitres": "ml",
    "liter": "l",
    "liters": "l",
    "litre": "l",
    "litres": "l",
    "cups": "cup",
    "tablespoon": "tbsp",
    "tablespoons": "tbsp",
    "teaspoon": "tsp",
    "teaspoons": "tsp",
    "slices": "slice",
    "pieces": "piece",
    "items": "item",
    "servings": "serving",
    "bars": "bar",
    "cans": "can",
    "bottles": "bottle",
    "bowls": "bowl",
    "handfuls": "handful",
    "scoops": "scoop",
    "glasses": "glass",
    "plates": "plate",
    "sandwiches": "sandwich",
    "tacos": "taco",
    "burritos": "burrito",
    "wraps": "wrap",
    "eggs": "egg",
    "muffins": "muffin",
    "cookies": "cookie",
    "bagels": "bagel",
    "nuggets": "nugget",
    "patties": "patty",
    "containers": "container",
    "pouches": "pouch",
}

MASS_G = {"g": 1.0, "kg": 1000.0, "oz": 28.349523125, "lb": 453.59237}
VOLUME_ML = {"ml": 1.0, "l": 1000.0, "cup": 240.0, "tbsp": 15.0, "tsp": 5.0, "fl oz": 29.5735295625}
FLOZ = VOLUME_ML["fl oz"]
NAMED_ML = {
    "short": 8 * FLOZ,
    "tall": 12 * FLOZ,
    "grande": 16 * FLOZ,
    "venti": 20 * FLOZ,
}
COUNT_UNITS = {
    "slice",
    "piece",
    "item",
    "each",
    "serving",
    "bar",
    "can",
    "bottle",
    "bowl",
    "handful",
    "scoop",
    "glass",
    "plate",
    "sandwich",
    "taco",
    "burrito",
    "wrap",
    "egg",
    "muffin",
    "cookie",
    "bagel",
    "nugget",
    "patty",
    "container",
    "pouch",
}
SIZE_FACTOR = {
    "extra small": 0.75,
    "small": 0.75,
    "medium": 1.0,
    "large": 1.25,
    "extra large": 1.25,
}


def canon_unit(unit: str | None) -> str:
    if not unit:
        return ""
    s = re.sub(r"\s+", " ", unit.lower().replace(".", "")).strip()
    if s in {"fl oz", "floz", "fluid ounce", "fluid ounces"}:
        return "fl oz"
    return UNIT_ALIAS.get(s, s)


def _parse_qty(raw: str) -> float | None:
    t = raw.strip()
    if re.fullmatch(r"\d+/\d+", t):
        a, b = t.split("/")
        return float(a) / float(b) if float(b) else None
    try:
        n = float(t)
    except ValueError:
        return None
    return n if n > 0 else None


def parse_household(label: str) -> dict | None:
    text = re.sub(r"\s+", " ", label).strip()
    m = re.match(r"^([\d.]+|\d+/\d+)\s+(.+)$", text, re.I)
    if not m:
        return None
    qty = _parse_qty(m.group(1))
    if qty is None:
        return None
    rest = m.group(2).lower()
    if rest.startswith("fl oz") or rest.startswith("fluid oz"):
        return {"qty": qty, "unit": "fl oz", "ml": qty * VOLUME_ML["fl oz"], "massG": None}
    first = re.split(r"[,/]", rest)[0].strip()
    unit = canon_unit(first.split()[0] if first else rest)
    if not unit:
        return None
    if unit in MASS_G:
        return {"qty": qty, "unit": unit, "ml": None, "massG": qty * MASS_G[unit]}
    if unit in VOLUME_ML:
        return {"qty": qty, "unit": unit, "ml": qty * VOLUME_ML[unit], "massG": None}
    return {"qty": qty, "unit": unit, "ml": None, "massG": None}


def scale_nutrition(food: dict, grams: float) -> dict:
    f = grams / 100.0
    return {
        "kcal": int(round(float(food["kcal"]) * f)),
        "protein": round(float(food.get("protein") or 0) * f, 1),
        "carbs": round(float(food.get("carbs") or 0) * f, 1),
        "fat": round(float(food.get("fat") or 0) * f, 1),
    }


def _pack(food: dict, grams: float, method: str, detail: str) -> dict:
    g = max(0.1, grams)
    return {"grams": g, "method": method, "detail": detail, **scale_nutrition(food, g)}


def _density(food: dict) -> float | None:
    house = parse_household(food.get("serveLabel") or "")
    serve_g = float(food.get("serveG") or 0)
    if house and house.get("ml") and house["ml"] > 0 and serve_g > 0:
        return serve_g / house["ml"]
    return None


def _volume_grams(food: dict, user_ml: float, named: bool) -> dict:
    density = _density(food)
    house = parse_household(food.get("serveLabel") or "")
    serve_g = float(food.get("serveG") or 0)
    if density is not None and house and house.get("ml"):
        grams = serve_g * (user_ml / house["ml"])
        method = "named-size+usda-density" if named else "fda-volume+usda-density"
        return _pack(
            food,
            grams,
            method,
            f"{user_ml:.0f} mL × USDA {food.get('serveLabel')} ({serve_g} g / {house['ml']:.0f} mL)",
        )
    return _pack(food, user_ml, "fda-volume+water", f"{user_ml:.0f} mL at 1 g/mL")


def _household_count(food: dict, user_qty: float, unit: str) -> float:
    house = parse_household(food.get("serveLabel") or "")
    serve_g = float(food.get("serveG") or 0)
    label = (food.get("serveLabel") or "").lower()
    if house and (house["unit"] == unit or unit in label):
        return serve_g * (user_qty / house["qty"])
    return serve_g * user_qty


def convert_portion(food: dict, qty: float, unit: str | None) -> dict:
    qty = qty if qty and qty > 0 else 1.0
    unit = canon_unit(unit)
    serve_g = float(food.get("serveG") or 0)
    label = food.get("serveLabel") or ""

    if unit in MASS_G:
        grams = MASS_G[unit] * qty
        return _pack(food, grams, "nist-mass", f"{qty} {unit} × NIST {MASS_G[unit]} g")

    if unit in NAMED_ML:
        return _volume_grams(food, NAMED_ML[unit] * qty, True)

    if unit in VOLUME_ML:
        house = parse_household(label)
        if house and (house["unit"] == unit or (house.get("ml") and canon_unit(house["unit"]) == unit)):
            return _pack(
                food,
                serve_g * (qty / house["qty"]),
                "usda-household",
                f"{qty} {unit} from USDA {label} = {serve_g} g",
            )
        if house and house.get("ml"):
            return _volume_grams(food, VOLUME_ML[unit] * qty, False)
        if unit in label.lower() or (unit == "cup" and re.search(r"\bcups?\b", label.lower())):
            return _pack(food, serve_g * qty, "usda-household", f"{qty} × USDA {label}")
        return _volume_grams(food, VOLUME_ML[unit] * qty, False)

    if not unit:
        return _pack(food, serve_g * qty, "usda-serving", f"{qty} × USDA {label} ({serve_g} g)")

    if unit in SIZE_FACTOR:
        house = parse_household(label)
        if house and (house["unit"] == unit or re.search(r"medium|small|large", label.lower())):
            base = serve_g * (qty / house["qty"])
        else:
            base = serve_g * qty
        return _pack(
            food,
            base * SIZE_FACTOR[unit],
            "usda-household",
            f"{qty} {unit} × USDA {label}",
        )

    if unit in COUNT_UNITS:
        return _pack(
            food,
            _household_count(food, qty, unit),
            "usda-household",
            f"{qty} {unit} × USDA {label} ({serve_g} g)",
        )

    blob = f"{label} {food.get('name') or ''}".lower()
    if unit in blob:
        return _pack(food, serve_g * qty, "usda-household", f"{qty} × USDA {label}")
    return _pack(food, serve_g * qty, "usda-serving", f"{qty} {unit} → USDA serving {serve_g} g")


def portion_tool_line(food: dict, qty: float, unit: str | None) -> str:
    result = convert_portion(food, qty, unit)
    u = unit or "serving"
    grams = int(round(result["grams"]))
    serve_g = int(round(float(food.get("serveG") or 0)))
    return (
        f"USDA {food.get('serveLabel')} ({serve_g} g) · "
        f"convert_portion {qty} {u} → {grams} g, {result['kcal']} kcal"
    )

File: src/lib/portions.ts
import type { ExtractedItem, Food, LogEntry } from '../types'

/**
 * Host-side portion tools. The VLM names foods and household units;
 * these functions turn units into grams and USDA calories.
 *
 * Sources:
 * - Mass: NIST avoirdupois (1 oz = 28.349523125 g, 1 lb = 453.59237 g)
 * - Volume: FDA 21 CFR 101.9 nutrition labeling (1 cup = 240 mL, 1 tbsp = 15 mL, 1 tsp = 5 mL)
 * - US fluid ounce: NIST (29.5735295625 mL)
 * - Household grams: the USDA row's serveG / serveLabel (FNDDS, SR, Foundation, Branded)
 * - Named drink sizes: published cup volumes (grande = 16 fl oz), then USDA density
 * - Calories: always food.kcal * grams / 100 from that USDA row
 */

export type PortionMethod =
  | 'nist-mass'
  | 'usda-household'
  | 'fda-volume+usda-density'
  | 'fda-volume+water'
  | 'named-size+usda-density'
  | 'usda-whole-item'
  | 'usda-serving'

export type PortionResult = {
  grams: number
  kcal: number
  protein: number
  carbs: number
  fat: number
  method: PortionMethod
  detail: string
}

const UNIT_ALIAS: Record<string, string> = {
  gram: 'g',
  grams: 'g',
  kilogram: 'kg',
  kilograms: 'kg',
  ounce: 'oz',
  ounces: 'oz',
  pound: 'lb',
  pounds: 'lb',
  lbs: 'lb',
  milliliter: 'ml',
  milliliters: 'ml',
  millilitre: 'ml',
  millilitres: 'ml',
  liter: 'l',
  liters: 'l',
  litre: 'l',
  litres: 'l',
  cups: 'cup',
  tablespoon: 'tbsp',
  tablespoons: 'tbsp',
  teaspoon: 'tsp',
  teaspoons: 'tsp',
  slices: 'slice',
  pieces: 'piece',
  items: 'item',
  servings: 'serving',
  bars: 'bar',
  cans: 'can',
  bottles: 'bottle',
  bowls: 'bowl',
  handfuls: 'handful',
  scoops: 'scoop',
  glasses: 'glass',
  plates: 'plate',
  sandwiches: 'sandwich',
  tacos: 'taco',
  burritos: 'burrito',
  wraps: 'wrap',
  eggs: 'egg',
  muffins: 'muffin',
  cookies: 'cookie',
  bagels: 'bagel',
  nuggets: 'nugget',
  patties: 'patty',
  containers: 'container',
  pouches: 'pouch',
}

/** NIST Handbook 44 avoirdupois / SI. */
const MASS_G: Record<string, number> = {
  g: 1,
  kg: 1000,
  oz: 28.349523125,
  lb: 453.59237,
}

/** FDA 21 CFR 101.9(b) labeling volumes, plus NIST fluid ounce. */
const VOLUME_ML: Record<string, number> = {
  ml: 1,
  l: 1000,
  cup: 240,
  tbsp: 15,
  tsp: 5,
  'fl oz': 29.5735295625,
}

/** Published beverage cups, expressed as FDA/NIST milliliters. */
const NAMED_ML: Record<string, { ml: number; detail: string }> = {
  short: { ml: 8 * 29.5735295625, detail: 'short 8 fl oz' },
  tall: { ml: 12 * 29.5735295625, detail: 'tall 12 fl oz' },
  grande: { ml: 16 * 29.5735295625, detail: 'grande 16 fl oz' },
  venti: { ml: 20 * 29.5735295625, detail: 'hot venti 20 fl oz' },
}

const COUNT_UNITS = new Set([
  'slice',
  'piece',
  'item',
  'each',
  'serving',
  'bar',
  'can',
  'bottle',
  'bowl',
  'handful',
  'scoop',
  'glass',
  'plate',
  'sandwich',
  'taco',
  'burrito',
  'wrap',
  'egg',
  'muffin',
  'cookie',
  'bagel',
  'nugget',
  'patty',
  'container',
  'pouch',
])

const SIZE_FACTOR: Record<string, number> = {
  'extra small': 0.75,
  small: 0.75,
  medium: 1,
  large: 1.25,
  'extra large': 1.25,
}

export function canonUnit(unit: string | null | undefined): string {
  if (!unit) return ''
  const s = unit.toLowerCase().replace(/\./g, '').replace(/\s+/g, ' ').trim()
  if (s === 'fl oz' || s === 'floz' || s === 'fluid ounce' || s === 'fluid ounces') return 'fl oz'
  return UNIT_ALIAS[s] ?? s
}

function parseQty(raw: string): number | null {
  const t = raw.trim()
  if (/^\d+\/\d+$/.test(t)) {
    const [a, b] = t.split('/').map(Number)
    return b ? a / b : null
  }
  const n = Number(t)
  return Number.isFinite(n) && n > 0 ? n : null
}

export type Household = {
  qty: number
  unit: string
  ml: number | null
  massG: number | null
}

/** Read the USDA household serving printed on the row. */
export function parseHousehold(label: string): Household | null {
  const text = label.replace(/\s+/g, ' ').trim()
  const m = text.match(/^([\d.]+|\d+\/\d+)\s+(.+)$/i)
  if (!m) return null
  const qty = parseQty(m[1])
  if (qty == null) return null
  let rest = m[2].toLowerCase()
  if (rest.startsWith('fl oz') || rest.startsWith('fluid oz')) {
    return { qty, unit: 'fl oz', ml: qty * VOLUME_ML['fl oz'], massG: null }
  }
  const first = rest.split(/[,/]/)[0]?.trim() ?? rest
  const unit = canonUnit(first.split(/\s+/)[0] ?? first)
  if (!unit) return null
  if (unit in MASS_G) return { qty, unit, ml: null, massG: qty * MASS_G[unit] }
  if (unit in VOLUME_ML) return { qty, unit, ml: qty * VOLUME_ML[unit], massG: null }
  return { qty, unit, ml: null, massG: null }
}

export function scaleNutrition(
  food: Pick<Food, 'kcal' | 'protein' | 'carbs' | 'fat'>,
  grams: number,
): Pick<LogEntry, 'kcal' | 'protein' | 'carbs' | 'fat'> {
  const f = grams / 100
  return {
    kcal: Math.round(food.kcal * f),
    protein: Math.round(food.protein * f * 10) / 10,
    carbs: Math.round(food.carbs * f * 10) / 10,
    fat: Math.round(food.fat * f * 10) / 10,
  }
}

function pack(food: Food, grams: number, method: PortionMethod, detail: string): PortionResult {
  const g = Math.max(0.1, grams)
  return { grams: g, method, detail, ...scaleNutrition(food, g) }
}

function densityGPerMl(food: Food): number | null {
  const house = parseHousehold(food.serveLabel)
  if (house?.ml && house.ml > 0 && food.serveG > 0) return food.serveG / house.ml
  return null
}

function volumeGrams(food: Food, userMl: number, named: boolean): PortionResult {
  const density = densityGPerMl(food)
  const house = parseHousehold(food.serveLabel)
  if (density != null && house?.ml) {
    const grams = food.serveG * (userMl / house.ml)
    return pack(
      food,
      grams,
      named ? 'named-size+usda-density' : 'fda-volume+usda-density',
      `${userMl.toFixed(0)} mL × USDA ${food.serveLabel} (${food.serveG} g / ${house.ml.toFixed(0)} mL)`,
    )
  }
  return pack(
    food,
    userMl,
    'fda-volume+water',
    `${userMl.toFixed(0)} mL at 1 g/mL (no USDA volume serving on this row)`,
  )
}

function householdCount(food: Food, userQty: number, unit: string): number {
  const house = parseHousehold(food.serveLabel)
  if (house && (house.unit === unit || food.serveLabel.toLowerCase().includes(unit))) {
    return food.serveG * (userQty / house.qty)
  }
  return food.serveG * userQty
}

/**
 * convert_portion(food, quantity, unit)
 * Grams from NIST/FDA/USDA. Calories from USDA per 100 g only.
 */
export function convertPortion(
  food: Food,
  item: Pick<ExtractedItem, 'quantity' | 'unit'>,
  opts?: { wholeProduceGrams?: number },
): PortionResult {
  const qty = Number.isFinite(item.quantity) && item.quantity > 0 ? item.quantity : 1
  const unit = canonUnit(item.unit)
  const whole = opts?.wholeProduceGrams

  if (unit in MASS_G) {
    const grams = MASS_G[unit] * qty
    return pack(food, grams, 'nist-mass', `${qty} ${unit} × NIST ${MASS_G[unit]} g`)
  }

  if (unit in NAMED_ML) {
    return volumeGrams(food, NAMED_ML[unit].ml * qty, true)
  }

  if (unit in VOLUME_ML) {
    const house = parseHousehold(food.serveLabel)
    if (house?.unit === unit || (house?.ml && canonUnit(house.unit) === unit)) {
      return pack(
        food,
        food.serveG * (qty / house.qty),
        'usda-household',
        `${qty} ${unit} from USDA ${food.serveLabel} = ${food.serveG} g`,
      )
    }
    if (house?.ml) {
      return volumeGrams(food, VOLUME_ML[unit] * qty, false)
    }
    const label = food.serveLabel.toLowerCase()
    if (label.includes(unit) || (unit === 'cup' && /\bcups?\b/.test(label))) {
      return pack(food, food.serveG * qty, 'usda-household', `${qty} × USDA ${food.serveLabel}`)
    }
    return volumeGrams(food, VOLUME_ML[unit] * qty, false)
  }

  if (whole != null && (!unit || unit in SIZE_FACTOR || COUNT_UNITS.has(unit) || unit === 'whole')) {
    const factor = SIZE_FACTOR[unit] ?? 1
    return pack(
      food,
      whole * qty * factor,
      'usda-whole-item',
      `USDA medium-item weight ${whole} g (row is a ${food.serveG} g garnish slice)`,
    )
  }

  if (!unit) {
    return pack(food, food.serveG * qty, 'usda-serving', `${qty} × USDA ${food.serveLabel} (${food.serveG} g)`)
  }

  if (unit in SIZE_FACTOR) {
    const house = parseHousehold(food.serveLabel)
    const base =
      house && (house.unit === unit || /medium|small|large/.test(food.serveLabel.toLowerCase()))
        ? food.serveG * (qty / house.qty)
        : food.serveG * qty
    return pack(food, base * (SIZE_FACTOR[unit] ?? 1), 'usda-household', `${qty} ${unit} × USDA ${food.serveLabel}`)
  }

  if (COUNT_UNITS.has(unit)) {
    return pack(
      food,
      householdCount(food, qty, unit),
      'usda-household',
      `${qty} ${unit} × USDA ${food.serveLabel} (${food.serveG} g)`,
    )
  }

  const label = `${food.serveLabel} ${food.name}`.toLowerCase()
  if (label.includes(unit)) {
    return pack(food, food.serveG * qty, 'usda-household', `${qty} × USDA ${food.serveLabel}`)
  }

  return pack(food, food.serveG * qty, 'usda-serving', `${qty} ${unit} → USDA serving ${food.serveG} g`)
}

/** One line for the VLM: USDA reference plus convert_portion for this item. */
export function portionToolLine(food: Food, item: Pick<ExtractedItem, 'quantity' | 'unit'>, result: PortionResult): string {
  const qty = item.quantity
  const unit = item.unit ?? 'serving'
  const grams = Math.round(result.grams)
  return `USDA ${food.serveLabel} (${Math.round(food.serveG)} g) · convert_portion ${qty} ${unit} → ${grams} g, ${result.kcal} kcal`
}
